casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala

```diff
@@ -96,50 +91,32 @@ object DagOperations {
+    * Conceptually, the LCA is the lowest point at which the histories of b1 and b2 diverge.
+    * We compute by finding the first block that is the "oldest" (has highest blocknum) common block.
     */
-  def greatestCommonAncestorF[F[_]: Monad: BlockStore](
-      b1: BlockMessage,
-      b2: BlockMessage,
-      genesis: BlockMessage,
+  def lowestCommonAncestorF[F[_]: Monad](
+      b1: BlockMetadata,
+      b2: BlockMetadata,
       dag: BlockDagRepresentation[F]
-  ): F[BlockMessage] =
+  ): F[BlockMetadata] = {
+    type BlockWithDepth = (Long, BlockMetadata)
+    implicit val blockNumDecreasing: Ordering[BlockWithDepth] =
+      (l: BlockWithDepth, r: BlockWithDepth) => r._1.compare(l._1)
+
+    def getParents(p: BlockMetadata): F[Set[BlockMetadata]] =
+      p.parents.map(v => dag.lookup(v)).sequence.map(_.toSet.flatten)
+
+    def extractParentsFromHighestNumBlock(s: SortedSet[BlockWithDepth]): F[SortedSet[BlockWithDepth]] = {
+      val (head, tail) = (s.head._2, s.tail)
+      getParents(head).map(nb => tail ++ nb.map(v => (v.blockNum, v)))
+    }
+
     if (b1 == b2) {
       b1.pure[F]
     } else {
-      def commonAncestorChild(
-          b: BlockMessage,
-          commonAncestors: Set[BlockMessage]
-      ): F[List[BlockMessage]] =
-        for {
-          childrenHashesOpt      <- dag.children(b.blockHash)
-          childrenHashes         = childrenHashesOpt.getOrElse(Set.empty[BlockHash])
-          children               <- childrenHashes.toList.traverse(ProtoUtil.unsafeGetBlock[F])
-          commonAncestorChildren = children.filter(commonAncestors)
-        } yield commonAncestorChildren
-
-      for {
-        b1Ancestors     <- bfTraverseF[F, BlockMessage](List(b1))(ProtoUtil.unsafeGetParents[F]).toSet
-        b2Ancestors     <- bfTraverseF[F, BlockMessage](List(b2))(ProtoUtil.unsafeGetParents[F]).toSet
-        commonAncestors = b1Ancestors.intersect(b2Ancestors)
-        gca <- bfTraverseF[F, BlockMessage](List(genesis))(commonAncestorChild(_, commonAncestors))
-                .findF(
-                  b =>
-                    for {
-                      childrenOpt <- dag.children(b.blockHash)
-                      children    = childrenOpt.getOrElse(Set.empty[BlockHash]).toList
-                      result <- children.existsM(
-                                 hash =>
-                                   for {
-                                     c <- ProtoUtil.unsafeGetBlock[F](hash)
-                                   } yield b1Ancestors(c) ^ b2Ancestors(c)
-                               )
-                    } yield result
-                )
-      } yield gca.get
+      val start = SortedSet.empty[BlockWithDepth] + ((b1.blockNum, b1)) + ((b2.blockNum, b2))
+      val result = Monad[F].iterateWhileM(start)(extractParentsFromHighestNumBlock)(_.size != 1).map(v => v.head)
+      result.map(_._2)
```

Comment:
the two consecutive maps and the result intermediate value can be eliminated without loss of readability.

Also it's more readable if written with one function call per line, like this:
```scala
Monad[F].iterateWhileM(start)(extractParentsFromHighestNumBlock)(_.size != 1)
       .map(v => v.head._2)
```